packages/core/collections/blob-dictionary.ts

```diff
@@ -0,0 +1,484 @@
+import { BytesBlob } from "@typeberry/bytes";
+import type { Comparator } from "@typeberry/ordering";
+import { asOpaqueType, assertNever, check, type Opaque, WithDebug } from "@typeberry/utils";
+
+const CHUNK_SIZE = 6;
+type CHUNK_SIZE = typeof CHUNK_SIZE;
+
+/**
+ * A function to transform a bytes chunk (up to 6 bytes into U48 number)
+ *
+ * Note that it uses 3 additional bits to store length(`value * 8 + len;`),
+ * It is needed to distinguish shorter chunks that have 0s at the end, for example: [1, 2] and [1, 2, 0]
+ * */
+export function bytesAsU48(bytes: Uint8Array): number {
+  const len = bytes.length;
+
+  check`${len <= CHUNK_SIZE} Length has to be <= ${CHUNK_SIZE}, got: ${len}`;
+
+  let value = 0;
+
+  for (let i = 0; i < len; i++) {
+    value = value * 256 + bytes[i];
+  }
+
+  return value * 8 + len;
+}
+
+type KeyChunk = Opaque<BytesBlob, `up to ${CHUNK_SIZE} bytes`>;
+type U48 = number;
+type SubKey<_K extends BytesBlob> = BytesBlob;
+type OriginalKeyRef<K> = K;
+
+type Leaf<K extends BytesBlob, V> = {
+  key: OriginalKeyRef<K>;
+  value: V;
+};
+
+class Node<K extends BytesBlob, V, C = MapChildren<K, V> | ListChildren<K, V>> {
```

Comment:
Just a personal preference, but I'd rather see the `BlobDictionary` class at the beginning of the file and the auxiliary classes at the end.